### PR TITLE
feat: Introduce opt-in deterministic relay data hashes (again)

### PR DESCRIPTION
**Pull request by mrice32**
Note: this should remain unmerged, as it will be audited separately from the rest of the codebase. The diff will be larger once this revert PR is merged: https://github.com/across-protocol/contracts/pull/638.

Original PR is here: https://github.com/across-protocol/contracts/pull/583.

This PR is identical to #583, except that it adds this deterministic feature to the 7683 implementation.

### CODE REVIEW

**Review by nicholaspai — APPROVED** [review]

**Comment by nicholaspai** [comment]
I've merged in https://github.com/across-protocol/contracts/pull/649 into this branch

File: contracts/SpokePool.sol [hunk 1]
```diff
@@ -470,9 +473,17 @@ abstract contract SpokePool is
      * @param fillDeadline The deadline for the relayer to fill the deposit. After this destination chain timestamp,
      * the fill will revert on the destination chain. Must be set between [currentTime, currentTime + fillDeadlineBuffer]
      * where currentTime is block.timestamp on this chain or this transaction will revert.
-     * @param exclusivityPeriod Added to the current time to set the exclusive relayer deadline,
-     * which is the deadline for the exclusiveRelayer to fill the deposit. After this destination chain timestamp,
-     * anyone can fill the deposit.
+     * @param exclusivityParameter This value is used to set the exclusivity deadline timestamp in the emitted deposit
+     * event. Before this destinationchain timestamp, only the exclusiveRelayer (if set to a non-zero address),
```
Comment (chrismaree) [inline contracts/SpokePool.sol:477]:
```suggestion
     * event. Before this destination chain timestamp, only the exclusiveRelayer (if set to a non-zero address),
```

File: contracts/SpokePool.sol [hunk 2]
```diff
@@ -487,62 +498,95 @@ abstract contract SpokePool is
...
+            depositor,
+            recipient,
             inputToken,
             outputToken,
             inputAmount,
             outputAmount,
             destinationChainId,
+            exclusiveRelayer,
             // Increment count of deposits so that deposit ID for this spoke pool is unique.
+            // @dev Implicitly casts from uint32 to uint256 by padding the left-most bytes with zeros. Guarantees
```
Comment (chrismaree) [inline contracts/SpokePool.sol:514]: this comment seems misplaced.

File: contracts/SpokePool.sol [hunk 3]
```diff
@@ -487,62 +498,95 @@ abstract contract SpokePool is
...
+            // that the 24 most significant bytes are 0.
             numberOfDeposits++,
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/SpokePool.sol]: see https://github.com/across-protocol/contracts/pull/639/commits/d3de7e1a5ced192181d52e1502371a81a169b627 for updated comments

File: contracts/SpokePool.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/SpokePool.sol]: https://github.com/across-protocol/contracts/pull/639/commits/d3de7e1a5ced192181d52e1502371a81a169b627

**Review by md0x — APPROVED** [review]
Looks good!

**Review by chrismaree — APPROVED** [review]

**Pull request merged**